proposals/0000-componentized-cabal.rst

```diff
@@ -0,0 +1,381 @@
+pre-existing packages from the package database or planned packages that
+need to be built: planned packages are identified only by a package
+identifier, and also store other solver information such as the flag
+assignment that was picked. These solver plan is the converted into an
+install plan in a one-to-one correspondence (mapping package identifiers
+into installed package identifiers).
+
+When setup dependencies were added to cabal-install, cabal-install's
+dependency solver was also refined to keep track of dependencies on a
+per-component basis (``ComponentDeps``).  Thus, although the graph of
+resolver packages is determined by the full, combined dependencies of
+all components in a package, it's also possible to determine the
+dependencies for a specific component. At the moment, this is only used
+to build Setup scripts (which constitute a component of its own) with
+only the necessary packages.
+
+**The proposal.** Dependency solving naturally is a package-level
+affair, since versions are ascribed to packages, not components.
+However, we would like build components individually: thus, we need to
+expand a package-based solver plan into a component-based install plan.
+This is how it works:
+
+1. For each package, we resolve the conditionals (using the solver
+   provided flag assignments) in order to produce a
+   ``PackageDescription``, which is essentially a Cabal file with all
+   conditionals stripped off. Prior to this step, we don't even know
+   what the components of the package necessarily will be (a component
+   can be marked as un-buildable through the dependency solving
+   process).
+
+2. We can now convert a graph of packages into a graph of components. In
+   doing so, the ``ComponentDeps`` tree of dependencies gets exploded into
+   an individual set of package level dependencies. Each component
+   identity is defined by looking at the input dependencies *of the
+   component*, as well other options which would affect the build. Note
+   that in a solver plan, the dependencies refer to *packages*, not
+   components: these package references are resolved to the *public
+   library component* of a package in question. After componentization,
+   ``ComponentDeps`` is eliminated from the install plan.
+
+3. A component is built by configuring its member package for solely the
+   component in question and then building it in a distinct build
+   directory. Execution of the install plan can now be parallelized over
+   components, rather than packages.
+
+**Determining if non-libraries are already installed.** In general, we
+can determine if a library is already installed by consulting the
+installed package database; however, no such database exists for
+executables. However, in ``new-build``, executables are installed to a
+deterministic directory in the Nix store; thus, we can simply check if
+the directory already exists in order to determine if an executable has
+already been built.
```

Comment:
Dir entry is insufficient for libraries since we need to load the `InstalledPackageInfo` (even if it is not used now, it is definitely used by Backpack).